### PR TITLE
test: parameterize SotW/delta in integration tests

### PR DESCRIPTION
**Pull request by fredlas**
Introduce a unified "generic" version of compareDiscoveryRequest and sendDiscoveryResponse for use in integration tests, allowing parameterization of SotW/delta. #4991 

Risk Level: none, test only
Testing: modified cds_ and ads_integration_test, integration/integration.cc.

### CODE REVIEW

**Comment by fredlas** [comment]
I thought a bit about some way to have these "generic" versions be expressed by the user in either just SotW style or just delta style, but it really didn't seem feasible. It would have taken a whole bunch of extra machinery, and constantly reaching into and messing with Any blobs. So, instead, when writing a compareDiscoveryRequest / sendDiscoveryResponse, the user should specify both what the state of the world should look like, along with what the additions/removals are.

File: test/common/grpc/grpc_client_integration.h [hunk 1]
```diff
@@ -52,21 +52,53 @@ class GrpcClientIntegrationParamTest
...
+    : public testing::TestWithParam<
+          std::tuple<Network::Address::IpVersion, ClientType, SotwOrDelta>> {
+public:
+  ~DeltaSotwIntegrationParamTest() override = default;
+  static std::string
+  protocolTestParamsToString(const ::testing::TestParamInfo<
+                             std::tuple<Network::Address::IpVersion, ClientType, SotwOrDelta>>& p) {
+    return absl::StrCat(
+        (std::get<0>(p.param) == Network::Address::IpVersion::v4 ? "IPv4_" : "IPv6_"),
+        (std::get<1>(p.param) == ClientType::GoogleGrpc ? "GoogleGrpc" : "EnvoyGrpc"),
```
Comment (htuch) [inline test/common/grpc/grpc_client_integration.h]: Nit: another `_` to separate gRPC and API type? Might be clearer actually via a `fmt::format`, but don't mind either way.

File: test/integration/cds_integration_test.cc [hunk 2]
```diff
@@ -32,24 +32,32 @@ const char ClusterName2[] = "cluster_2";
...
       : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, ipVersion(),
-                            ConfigHelper::discoveredClustersBootstrap("GRPC")) {}
+                            ConfigHelper::discoveredClustersBootstrap(
+                                sotwOrDelta() == SotwOrDelta::Sotw ? "GRPC" : "DELTA_GRPC")) {
+    sotw_or_delta_ = sotwOrDelta();
+  }
 
-  CdsIntegrationTest(Network::Address::IpVersion ip_version, const std::string& config)
-      : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, ip_version, config) {}
+  // TODO TODO TRIM?  CdsIntegrationTest(Network::Address::IpVersion ip_version, const std::string&
```
Comment (htuch) [inline test/integration/cds_integration_test.cc]: ?

File: test/common/grpc/grpc_client_integration.h (hunk 1 above)
Comment (fredlas) [reply test/common/grpc/grpc_client_integration.h]: Done.

File: test/integration/cds_integration_test.cc (hunk 2 above)
Comment (fredlas) [reply test/integration/cds_integration_test.cc]: Done.

**Comment by fredlas** [comment]
That would be ideal, but that's what my initial comment was about. I think it's too awkward to be worth it.

I've been starting to think that xDS resources ought to have a standardized "name" field, that they all have. Actually, like what delta has, now that the Resource proto has a name field. That's what would be needed to make the cleanup here reasonable.

**Comment by htuch** [comment]
@fredlas sure, I think this PR is a strict improvement on the existing situation in terms of unifying. There might be some further cleanup later, but happy to ship this.

**Pull request merged**